orttraining/orttraining/python/training/utils/hooks/_activation_summarizer.py

```diff
@@ -0,0 +1,86 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import os
+import warnings
+import torch
+
+from ._subscriber_manager import _ModuleHookSubscriberBase
+
+
+class StatisticsSubscriber(_ModuleHookSubscriberBase):
+    """
+    This subscriber is used to dump the activation's statistics to a file.
+
+    In the forward pass, summarize the tensor's statistics and write to a file.
+    In the backward pass, summarize the tensor's statistics and write into another file.
+
+    Each activation will be summarized into 1 or 2 files, depending on whether it is used in the backward pass.
+    So for each run step, there will be many files.
+
+    `merge_activation_summary.py` can be used to merge the files into one file per training step.
+    """
+
+    def __init__(self, output_dir, start_step_inclusive=0, end_step_exclusive=1000000, override_output_dir=False):
+        super().__init__(start_step=start_step_inclusive, end_step=end_step_exclusive)
+        self._output_dir = output_dir
+        if os.path.exists(self._output_dir):
+            if override_output_dir:
+                import shutil
+
+                warnings.warn(f"Output directory {self._output_dir} already exists, override it.")
+                shutil.rmtree(self._output_dir)
+            else:
+                raise RuntimeError(
+                    f"Output directory {self._output_dir} already exists. "
+                    "Set override_output_dir=True for StatisticsSubscriber if this is the intention."
+                )
+
+    def forward(self, activation_sequence_id, activation, depth, name, step):
+        output_file_path = os.path.join(f"{self._output_dir}", f"step_{step}")
+        return StatisticsSubscriber._summarize_activations(activation, depth, name + " forward run", output_file_path)
+
+    def backward(self, activation_sequence_id, activation, depth, name, step):
+        output_file_path = os.path.join(f"{self._output_dir}", f"step_{step}")
+        return StatisticsSubscriber._summarize_activations(activation, depth, name + " backward run", output_file_path)
+
+    @staticmethod
+    def _summarize_activations(tensor, depth, name, step_folder):
+        if tensor is None or not isinstance(tensor, torch.Tensor):
+            print(f"{name} not a torch tensor, value: {tensor}")
+            return None
+
+        d = f"{step_folder}"
+        isExist = os.path.exists(d)
+        if not isExist:
+            # Create a new directory because it does not exist
+            os.makedirs(d)
+
+        path = os.path.join(d, f"{name}")
```

Comment:
I see a lot of path manipulation in this file. Have you considered using pathlib? It is a robust way of working with paths: https://docs.python.org/3/library/pathlib.html